mgmt/updatehub, storage/flash_map: replace TinyCrypt by PSA

tomi-font opened the pull request:
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/72475 for the build failures (and the ROM/RAM usage measurements).

As part of ongoing work to move away from TinyCrypt and towards PSA (#43712), introduce a PSA option and remove the TinyCrypt one for the SHA-256 implementation.

The Mbed TLS implementations are modified to use `mbedtls_sha256` directly for smaller code size.

See the commit messages for more details.

ROM and RAM usage difference measurements below. The TinyCrypt version is used as reference, and all the measurements are relative to it.
They were obtained by compiling `samples/subsys/mgmt/updatehub` without additional overlay. 
Functions from both `subsys/mgmt/updatehub/updatehub_integrity.c` and `subsys/storage/flash_map/flash_map_integrity.c` are compiled in and used.
<br>
On `nrf52840dk/nrf52840`:
```
    | TC (before) | Mbed TLS (before) | Mbed TLS (after) | PSA (after) * |
----|-------------|-------------------|------------------|---------------|
ROM | reference   | +1288 B           | -40 B            | +616/728 B    |
RAM | reference   | +488 B            | -64 B            | +0 B          |
```
\* The increase in the PSA version is a one-time cost of enabling PSA in Mbed TLS which enables an entropy source (which itself has its own dependencies). The first measurement is with `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`, the second with `CONFIG_MBEDTLS_PSA_CRYPTO_LEGACY_RNG`. 
<br>
On `nrf5340dk/nrf5340/cpuapp/ns` (with `CONFIG_PSA_CRYPTO_ENABLE_ALL` additionally disabled which https://github.com/zephyrproject-rtos/zephyr/pull/73412 takes care of):
```
    | TC (before) | Mbed TLS (before) | Mbed TLS (after) | PSA (after) * |
----|-------------|-------------------|------------------|---------------|
ROM | reference   | +1272 B           | +248 B           | -736 B        |
RAM | reference   | +440 B            | -112 B           | -120 B        |
```
\* Expected decrease in the PSA version of an NS platform because the crypto implementation is already present in the TF-M image.

tomi-font committed:
mgmt: updatehub: replace TinyCrypt by PSA

As part of ongoing work to move away from TinyCrypt and towards PSA
(#43712), introduce a PSA option and remove the TinyCrypt one for the
SHA-256 implementation.

The Mbed TLS implementation is modified to use `mbedtls_sha256`
directly for smaller code size.

The reliance of mgmt/updatehub on storage/flash_map's configuration
(`FLASH_AREA_CHECK_INTEGRITY_BACKEND`) is removed.
The choice of which implementation to use is made automatically,
based on whether a PSA API provider is present (`PSA_CRYPTO_CLIENT`).

This commit also add a test case with PSA (based on Mbed TLS)
in samples/subsys/mgmt/updatehub/sample.yaml.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>
Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font committed:
storage: flash_map: replace TinyCrypt by PSA

As part of ongoing work to move away from TinyCrypt and towards PSA
(#43712), introduce a PSA option and remove the TinyCrypt one for the
SHA-256 implementation.

The Mbed TLS implementation is modified to use `mbedtls_sha256`
directly for smaller code size.

As of now the implementation defaults to PSA only if TF-M is enabled
because a dependency loop happens if using `PSA_CRYPTO_CLIENT` as a
condition in `FLASH_AREA_CHECK_INTEGRITY_BACKEND`.

A test case is added for the PSA implementation, and an NS platform is
added to the base test case to verify the compilation on a TF-M-enabled
platform.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

valeriosetti commented on subsys/mgmt/updatehub/updatehub_integrity.c:
```diff
@@ -18,43 +24,22 @@ int updatehub_integrity_init(struct updatehub_crypto_context *ctx)
-	ret = tc_sha256_init(&ctx->sha256sum);
-	if (ret != TC_CRYPTO_SUCCESS) {
-		LOG_DBG("Invalid integrity context");
+#if defined(CONFIG_MBEDTLS_PSA_CRYPTO_CLIENT)
```
I think that 
```suggestion
#if defined(CONFIG_PSA_CRYPTO_CLIENT)
```
would be better as you already did in `subsys/mgmt/updatehub/Kconfig`.
The same holds for other occurrences of the `CONFIG_MBEDTLS_PSA_CRYPTO_CLIENT`.

tomi-font replied:
True. That was an oversight; well spotted. I'll replace the occurrences in this file.

frkv reviewed (approved):
Happy about what I'm seeing here!

tomi-font commented:
Dropped the cherry pick of https://github.com/zephyrproject-rtos/zephyr/pull/72475 after having confirmed that CI is fully green with it (so the failures are expected and CI will just need to be re-triggered when the PR is merged) and added migration guide entries.

No more upcoming modifications; this is ready for review.

valeriosetti reviewed (dismissed):
LGTM (I'm assuming that the CI will be OK once https://github.com/zephyrproject-rtos/zephyr/pull/72475 is merged)

de-nordic commented on subsys/storage/flash_map/flash_map_integrity.c:
```diff
@@ -81,50 +71,47 @@ int flash_area_check_int_sha256(const struct flash_area *fa,
 	if (memcmp(hash, fac->match, SHA256_DIGEST_SIZE)) {
-#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_TC)
+#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_PSA)
+		/* The operation has already been terminated. */
```
Shouldn't this part also `goto error`?
There is `psa_has_abort()` that I think is supposed to be called on failure.

tomi-font replied:
No, it doesn't need to. If `psa_hash_finish()` returns successfully there is nothing more to do, and here it is the case (otherwise we will have `goto error` already).
This case is a failure in the flash area check (checksum mismatch), not in the PSA hash operation.

tomi-font commented:
> LGTM (I'm assuming that the CI will be OK once #72475 is merged)

Correct. Anyway, a PR with failing CI wouldn't be merged.

de-nordic reviewed (dismissed):
Flash Map integrity check looks ok.

ceolin commented on subsys/storage/flash_map/CMakeLists.txt:
```diff
@@ -6,6 +6,4 @@ zephyr_sources_ifdef(CONFIG_FLASH_MAP_SHELL flash_map_shell.c)
-if(CONFIG_FLASH_AREA_CHECK_INTEGRITY_MBEDTLS)
-  zephyr_library_link_libraries_ifdef(CONFIG_MBEDTLS mbedTLS)
-endif()
+zephyr_library_link_libraries_ifdef(CONFIG_MBEDTLS mbedTLS)
```
Shouldn't it be guarded by `CONFIG_FLASH_AREA_CHECK_INTEGRITY_MBEDTLS` ? Wondering if there is a case where Mbed TLs together with TF-M (and consequently `FLASH_AREA_CHECK_INTEGRITY_PSA`).

valeriosetti replied:
@tomi-font is on vacation so I'm inheriting this PR in the meantime.
I think he removed the check because otherwise in builds with TF-M there is this failure:
```
flash_map_integrity.c:24:10: fatal error: psa/crypto.h: No such file or directory
   24 | #include <psa/crypto.h>
```
It just needs to find the crypto header somewhere. The most "elegant" solution would be to specify:
- `mbedTLS` library when building/linking against Mbed TLS, or
- `<build-folder>/tfm/api_ns/interface/include/psa/crypto.h`

But here it comes my newbe question. In TF-M CMakeList file there is a variable named `TFM_INTERFACE_INCLUDE_DIR` which might do exactly what we're looking for, but that's not accessible from `zephyr/subsys/storage/flash_map/CMakeLists.txt`. Is there a way to make use of that to get what I described above?

valeriosetti commented:
This PR was rebased in order to re-run the CI, but code was not changed. Now the CI is fully green so I think the PR is ready for reviews again :)

nandojve commented:
Hi @valeriosetti ,

Thank you for introducing these changes. I made tests using mbedTLS and everything seems fine.

Now, I would like to test using PSA. Could you share the configuration necessary to enable PSA_CLIENT on nucleo_f767zi?

That configuration should be added as `samples/subsys/mgmt/updatehub/overlay-psa.conf` and another one (if is not the same) to build using TFM like `samples/subsys/mgmt/updatehub/overlay-tfm.conf`. In the same way, it is very important add the two configuration at `samples/subsys/mgmt/updatehub/samples.yaml` to build using those options. This way the sample may help to track changes on those modules since it uses a lot of dependencies involved.

valeriosetti commented:
@nandojve sorry, but I'm not sure I understood the point here.

> Now, I would like to test using PSA. Could you share the configuration necessary to enable PSA_CLIENT on nucleo_f767zi?

Is this a request for change for this PR? Can it be addressed in a follow up PR?

> and another one (if is not the same) to build using TFM

Maybe I'm missing something, but as far as I can tell STM32F767 does not have TrustZone and it's single core so it should not support support TF-M as well. Am I wrong?

nandojve commented:
> @nandojve sorry, but I'm not sure I understood the point here.
> 
> > Now, I would like to test using PSA. Could you share the configuration necessary to enable PSA_CLIENT on nucleo_f767zi?
> 
> Is this a request for change for this PR? Can it be addressed in a follow up PR?
> 
> > and another one (if is not the same) to build using TFM
> 
> Maybe I'm missing something, but as far as I can tell STM32F767 does not have TrustZone and it's single core so it should not support support TF-M as well. Am I wrong?

1- I added this example just to clarify something about PSA. I understood that PSA is an API and should be able to use it with/without TFM. Is this assumption wrong? For me it very strange to define a API that depends on TFM. I even understood that (is not the case, just as example) TinyCrypt could be executed behind PSA api. I've been waiting for this change since 2020 since it was mentioned on this proposal #29650 and I would like to understand if I need wait more.

2- Yes, it is necessary add the missing overlay-psa/tfm otherwise this work is incomplete. I already see that your @tomi-font already made it for `nrf5340dk/nrf5340/cpuapp/ns`. What I'm asking is to add the overlay file + to enable on CI.
I understand since this is a major goal it can be finish in the RC-1.

3- Just to mention that besides the #43249 be closed it still a valid use case.

I hope you can help to clarify above points. Maybe I still missing something here about this API change.

frkv commented:
> 1- I added this example just to clarify something about PSA. I understood that PSA is an API and should be able to use it with/without TFM. Is this assumption wrong? For me it very strange to define a API that depends on TFM. I even understood that (is not the case, just as example) TinyCrypt could be executed behind PSA api. I've been waiting for this change since 2020 since it was mentioned on this proposal #29650 and I would like to understand if I need wait more.

I will comment on this, as I am one of the persons that have been planning the effort of getting PSA crypto support in within the Zephyr LTS scope.

The reason we are using TF-M as a gating requirement (right now) is lack of a certain feature to whole-heartedly getting every board to support PSA crypto. There is a feature that is present in TF-M which is not available in non TF-M device. This is the lack of the PSA secure storage APIs in a usable version. TF-M has NVM aware implementation for persistent key storage (PSA ITS) and options for secured certificate storage (PSA PS). Devices without TF-M only has one option for this, and it is through the ITS support that is natively built into Mbed TLS (C File based).

Nordic Semiconductor is planning to add a new subystem/library called "Trusted Storage" which gives NVM-aware (encrypted) key storage to give Native ITS and PS support to devices that can't/won't use TF-M. Unfortunately, addition of such a component is not something that we though we were able to get in before the Zephyr LTS. This has been discussed both in Zephyr Security Committee and working group. Right now we are doing a "best effort" and in this case we are able to do proper PSA support for TF-M enabled devices (so we check things with BUILD_WITH_TFM) and we only give limited support for non TF-M devices (in case persistent key is not a direct requirement)

_Once we have Trusted Storage (or a variant of this) in Zephyr is when we can make the claim that PSA crypto is fully supported or supportable by any and all device types. Right now we are only doing an effort to transition as much as we can. We appologize that the support isn't there yet, and as you state, this has been a long running effort. We are doing as much as we can with the limitations we have, right now..._

frkv commented:
> 3- Just to mention that besides the #43249 be closed it still a valid use case.

MBEDTLS_ECP_C is still a a dependency for anything that depends on this, of course... But note that MBEDTLS_ECP_C is likely a fully internalized construct in the timeline of Mbed TLS 4.0.0. Zephyr is intentionally working towards PSA crypto APIs being the defacto crypto APIs in the system.

Note that ECP has been largely redesigned in Mbed TLS and the focus has been on front-end APIs (ECDH, ECDSA (+deterministic), EdDSA, x25519, ECJPAKE and SPAKE2+). If there are other types of low-level arithmetic being used (requiring direct access to ECP-functionality) the best bet that Zephyr can make is to ensure that these are hidden behind supported APIs and become a one-off. 

We have no guarantees that ECP layer will be a supported and usable API and it will for sure not be used in context where TF-M is enabled in the system. There you either need to extend the crypto toolbox with new features (according to protocol, algorithms) or you need to invest time to establish a custom Application Root of Trust.

_It is understood that this is an impactful change for different implementations, but unfortunately the transition towards this new APIs necessarily needs to be impactful for the proper establishment of key isolation (by opaque references). This is not possible without an API change_

valeriosetti commented:
@nandojve thanks for the reply. Here's my comment:

> That configuration should be added as samples/subsys/mgmt/updatehub/overlay-psa.conf and another one (if is not the same) to build using TFM like samples/subsys/mgmt/updatehub/overlay-tfm.conf

> Yes, it is necessary add the missing overlay-psa/tfm otherwise this work is incomplete.

I modified the first commit adding a new test case in `samples/subsys/mgmt/updatehub/sample.yaml`. This follows the very same pattern that @tomi-font already did for `tests/subsys/storage/flash_map/testcase.yaml`: it adds a test using PSA APIs. Test with TF-M is not supported by the `frdm_k64f`, but this is actually not really relevant: PSA is an interface for crypto APIs and it does not matter who provides them from the `updatehub/flash-map` point of view.
Wdyt?

> I understood that PSA is an API and should be able to use it with/without TFM. Is this assumption wrong?

Correct, PSA APIs is just an interface. So far in Zephyr it can be implemented/provided through Mbed TLS o TF-M (which itself uses Mbed TLS under the hood BTW).

> I even understood that (is not the case, just as example) TinyCrypt could be executed behind PSA api

No, as far as I know TinyCrypt does not have a PSA interface.

> Just to mention that besides the https://github.com/zephyrproject-rtos/zephyr/issues/43249 be closed it still a valid use case.

Please note that Mbed TLS greatly improved support for EC keys [recently](https://github.com/Mbed-TLS/mbedtls/releases) in PSA (after the PR you mentioned). Also Zephyr now has a better support for `PSA_WANT` and `USE_PSA` (#72243). So you might want to check support again and, if something is still missing, I can help with that.

nandojve reviewed (dismissed):
Hi @frkv and @valeriosetti ,

Thank you so much for your explanation. It helped me to understand more about this complex topic.

> I modified the first commit adding a new test case in samples/subsys/mgmt/updatehub/sample.yaml. This follows the very same pattern that @tomi-font already did for tests/subsys/storage/flash_map/testcase.yaml: it adds a test using PSA APIs. Test with TF-M is not supported by the frdm_k64f, but this is actually not really relevant: PSA is an interface for crypto APIs and it does not matter who provides them from the updatehub/flash-map point of view.
Wdyt?

We use the crypto not only over flash but we check the download content too. This helps us to isolate problems between networking/protocols and the file system.

In the future we intent to move the `platform_allow` from common to each specific verification to get it more flexible and cover TFM cases. It is not necessary at moment and I'm happy to have the `overlay-psa.conf` to make it explicit for users.

Thank you for that!
